### ISSUE TITLE
Feature: support crio+crun container runtime

### ISSUE THREAD

**Issue by willzhang** [labring/sealos#2084]
### What is the problem this feature will solve?

support more container runtime.

### If you have solution，please describe it

```
sealos run registry.cn-shenzhen.aliyuncs.com/cnmirror/kubernetes-crio:v1.25.3 --single
sealos run labring/helm:v3.10.1
sealos run labring/calico:v3.24.1
```

```
root@ubuntu:~# kubectl get nodes -o wide
NAME     STATUS   ROLES           AGE   VERSION   INTERNAL-IP     EXTERNAL-IP   OS-IMAGE           KERNEL-VERSION      CONTAINER-RUNTIME
ubuntu   Ready    control-plane   17m   v1.25.3   192.168.72.51   <none>        Ubuntu 22.04 LTS   5.15.0-27-generic   cri-o://1.25.1
root@ubuntu:~# 
```
coredns CrashLoopBackOff   
```
root@ubuntu:~# kubectl get pods -A
NAMESPACE          NAME                                       READY   STATUS             RESTARTS      AGE
calico-apiserver   calico-apiserver-7bddd86d98-dk4cm          1/1     Running            0             10m
calico-apiserver   calico-apiserver-7bddd86d98-q54lx          1/1     Running            0             10m
calico-system      calico-kube-controllers-85666c5b94-jl4cr   1/1     Running            0             12m
calico-system      calico-node-npht4                          1/1     Running            0             12m
calico-system      calico-typha-598b5559d-bhr2b               1/1     Running            0             12m
calico-system      csi-node-driver-sq95f                      2/2     Running            0             11m
kube-system        coredns-565d847f94-7s6gj                   0/1     CrashLoopBackOff   7 (16s ago)   16m
kube-system        coredns-565d847f94-jmcgm                   0/1     CrashLoopBackOff   7 (11s ago)   16m
kube-system        etcd-ubuntu                                1/1     Running            0             17m
kube-system        kube-apiserver-ubuntu                      1/1     Running            0             17m
kube-system        kube-controller-manager-ubuntu             1/1     Running            1 (14m ago)   17m
kube-system        kube-proxy-nz762                           1/1     Running            0             16m
kube-system        kube-scheduler-ubuntu                      1/1     Running            2 (14m ago)   17m
tigera-operator    tigera-operator-6675dc47f4-cpwx5           1/1     Running            0             12m
```

logs
```
root@ubuntu:~# kubectl -n kube-system logs -f deploy/coredns
Found 2 pods, using pod/coredns-565d847f94-7s6gj
.:53
[INFO] plugin/reload: Running configuration SHA512 = 591cf328cccc12bc490481273e738df59329c62c0b729d94e8b61db9961c2fa5f046dd37f1cf888b953814040d180f52594972691cd6ff41be96639138a43908
CoreDNS-1.9.3
linux/amd64, go1.18.2, 45b0a11
[FATAL] plugin/loop: Loop (127.0.0.1:45745 -> :53) detected for zone ".", see https://coredns.io/plugins/loop#troubleshooting. Query: "HINFO 5242585425827747503.2346149046908645429."
root@ubuntu:~# 
```
/etc/resolv.conf config

```
root@ubuntu:~# cat /etc/resolv.conf 
# This is /run/systemd/resolve/stub-resolv.conf managed by man:systemd-resolved(8).
# Do not edit.
#
# This file might be symlinked as /etc/resolv.conf. If you're looking at
# /etc/resolv.conf and seeing this text, you have followed the symlink.
#
# This is a dynamic resolv.conf file for connecting local clients to the
# internal DNS stub resolver of systemd-resolved. This file lists all
# configured search domains.
#
# Run "resolvectl status" to see details about the uplink DNS servers
# currently in use.
#
# Third party programs should typically not access this file directly, but only
# through the symlink at /etc/resolv.conf. To manage man:resolv.conf(5) in a
# different way, replace this symlink by a static file or a different symlink.
#
# See man:systemd-resolved.service(8) for details about the supported modes of
# operation for /etc/resolv.conf.

nameserver 127.0.0.53
options edns0 trust-ad
search .
```

i must disable 127.0.0.53 for ubuntu default dns
```
mkdir -p /etc/systemd/resolved.conf.d/
cat >/etc/systemd/resolved.conf.d/98-disable-127-53.conf << EOF
[Resolve]
DNSStubListener=no
EOF

systemctl daemon-reload
systemctl restart systemd-resolved.service
```
coredns will be Running, but containerd not need do this, why?
```
root@ubuntu:~# kubectl -n kube-system rollout restart deployment/coredns
deployment.apps/coredns restarted

root@ubuntu:~# kubectl -n kube-system get pods 
NAME                             READY   STATUS    RESTARTS      AGE
coredns-6bc9c8896c-4phff         1/1     Running   0             8s
coredns-6bc9c8896c-k5f4l         1/1     Running   0             7s
etcd-ubuntu                      1/1     Running   0             21m
kube-apiserver-ubuntu            1/1     Running   0             21m
kube-controller-manager-ubuntu   1/1     Running   1 (19m ago)   21m
kube-proxy-nz762                 1/1     Running   0             21m
kube-scheduler-ubuntu            1/1     Running   2 (18m ago)   21m
root@ubuntu:~# 
```

**Comment by willzhang** [comment]
or delete `#loop`
```
root@ubuntu:/etc/cni/net.d# kubectl -n kube-system edit configmap coredns
# Please edit the object below. Lines beginning with a '#' will be ignored,
# and an empty file will abort the edit. If an error occurs while saving this file will be
# reopened with the relevant failures.
#
apiVersion: v1
data:
  Corefile: |
    .:53 {
        errors
        health {
           lameduck 5s
        }
        ready
        kubernetes cluster.local in-addr.arpa ip6.arpa {
           pods insecure
           fallthrough in-addr.arpa ip6.arpa
           ttl 30
        }
        prometheus :9153
        forward . /etc/resolv.conf {
           max_concurrent 1000
        }
        cache 30
        #loop
        reload
        loadbalance
    }
kind: ConfigMap
metadata:
  creationTimestamp: "2022-11-11T03:17:38Z"
  name: coredns
  namespace: kube-system
  resourceVersion: "11307"
  uid: f959dd94-5f51-4374-90ff-c26e8eef17fb
```

**Comment by cuisongliu** [comment]
> fixed by
> 
> ```
> KUBELET_EXTRA_ARGS="--resolv-conf /run/systemd/resolve/resolv.conf 
> ```

maybe this is a  bug ,using sealos install 1.24.0 try it.

**Comment by cuisongliu** [comment]
https://github.com/labring/sealos/issues/2074

**Comment by cuisongliu** [comment]
<img width="1339" alt="image" src="https://user-images.githubusercontent.com/4150905/201298673-d3edfd73-fbf7-4074-8c26-2a8722432b47.png">
It seems that k8s has been processed, but it doesn't take effect?

**Comment by willzhang** [comment]
The RP: https://github.com/labring/cluster-image/pull/252

## have basic test success
```
root@node1:~# kubectl get nodes -o wide
NAME    STATUS   ROLES           AGE    VERSION   INTERNAL-IP     EXTERNAL-IP   OS-IMAGE           KERNEL-VERSION      CONTAINER-RUNTIME
node1   Ready    control-plane   167m   v1.25.3   192.168.72.50   <none>        Ubuntu 22.04 LTS   5.15.0-27-generic   cri-o://1.25.1
node2   Ready    <none>          166m   v1.25.3   192.168.72.51   <none>        Ubuntu 22.04 LTS   5.15.0-27-generic   cri-o://1.25.1
node3   Ready    <none>          166m   v1.25.3   192.168.72.52   <none>        Ubuntu 22.04 LTS   5.15.0-27-generic   cri-o://1.25.1
root@node1:~# kubectl get pods -A
NAMESPACE          NAME                                       READY   STATUS    RESTARTS       AGE
calico-apiserver   calico-apiserver-5bdbcd4467-gdfvd          1/1     Running   1              165m
calico-apiserver   calico-apiserver-5bdbcd4467-qtv5r          1/1     Running   1              165m
calico-system      calico-kube-controllers-85666c5b94-5zxmw   1/1     Running   1              166m
calico-system      calico-node-gkvsm                          1/1     Running   1              166m
calico-system      calico-node-xrzn8                          1/1     Running   1              166m
calico-system      calico-node-zk6kd                          1/1     Running   1              166m
calico-system      calico-typha-7dd4b4b4cf-c99n2              1/1     Running   1              166m
calico-system      calico-typha-7dd4b4b4cf-krfqj              1/1     Running   1              166m
calico-system      csi-node-driver-bktjd                      2/2     Running   2              165m
calico-system      csi-node-driver-k7hwj                      2/2     Running   2              165m
calico-system      csi-node-driver-mcswq                      2/2     Running   2              165m
kube-system        coredns-565d847f94-6zqd9                   1/1     Running   1              167m
kube-system        coredns-565d847f94-wshz6                   1/1     Running   1              167m
kube-system        etcd-node1                                 1/1     Running   1              167m
kube-system        kube-apiserver-node1                       1/1     Running   1              167m
kube-system        kube-controller-manager-node1              1/1     Running   4 (119m ago)   167m
kube-system        kube-proxy-8hvcp                           1/1     Running   1              167m
kube-system        kube-proxy-985td                           1/1     Running   1              167m
kube-system        kube-proxy-zwvn7                           1/1     Running   1              167m
kube-system        kube-scheduler-node1                       1/1     Running   4 (119m ago)   167m
kube-system        kube-sealos-lvscare-node2                  1/1     Running   1              166m
kube-system        kube-sealos-lvscare-node3                  1/1     Running   1              166m
tigera-operator    tigera-operator-6675dc47f4-f8wmt           1/1     Running   6 (119m ago)   166m
```
## but have two problems
1、crio access private registry use config.json, 

```
[crio.image]
global_auth_file = "/var/lib/kubelet/config.json"
```
now At present, this piece is dead with `YWRtaW46cGFzc3cwcmQ=`
```
root@ubuntu:/data/git/cluster-image/cri-o# cat etc/config.json.tmpl 
{
        "auths": {
                "{{ .registryDomain }}:{{ .registryPort }}": {
                        "auth": "YWRtaW46cGFzc3cwcmQ="
                }
        }
}
```

2、when i run .github/scripts/kubernetes.sh

**Comment by willzhang** [comment]
It use /etc/resolv.conf 

![image](https://user-images.githubusercontent.com/33590311/201453910-b032141f-3631-4b1a-bcc3-2fc60a6bb855.png)

**Comment by willzhang** [comment]
Tested with sealos v4.1.3 and v4.1.4 -rc1

- v4.1.3 ok with kubernetes v1.24.x v1.25.x
- v4.1.4-rc1 failed with kubernetes v1.24.x v1.25.x  (have resolvConf: /etc/resolv.conf in kubeadm-init.yaml )

**Comment by willzhang** [comment]
> 

fixed in v4.1.4-rc2

**Comment by willzhang** [comment]
https://github.com/labring/cluster-image/pull/252
